update_json for module intermediate comms

bcoca opened the pull request:
- new 'live' keyword to indicate you want intermediate (live) updates from module execution
- added 'live' intelligence to 'run_command' for those modules executing commands using it

TODO:
- displays when -vvvvv ... really should be using callback now that sivel added queuing

implements https://github.com/ansible/proposals/issues/92

bcoca commented:
@tsipinakis feel free to push to this PR, im currently dealing with other things but I def want to get back to this .. if someone else beats me too it, GREAT!

tsipinakis commented:
@bcoca I'm not a part of the org so you'll need to give me access to your fork for me to be able to update the PR. (Just tried and got access denied)

bcoca commented:
you dont need to be part of the org, github allows pushing to specific PRs by default .. but this is very old branch and might predate that ...

tsipinakis commented:
I think you're referring to the `Allow edits from maintainers` feature which allows pushes only to those that have write access to the original repo.
It'd probably be a nightmare to allow any user to push to PRs without any access control in place.

tsipinakis commented:
Whoops, that's what I get for not testing python3, pushed a fixup.

Perhaps it'd be better for me to open a new PR for this?

danieljaouen commented:
@tsipinakis What is the status of this pull request.  Interested because this will fix #29403.

tsipinakis commented:
@danieljaouen Paused until I get some more time to see why the tests are (still) failing. If you want to take it over go right ahead.

danieljaouen commented:
@tsipinakis Perhaps @bcoca could shed some light on why the tests are failing?

acozine commented:
@tsipinakis @danieljaouen this PR needs to be rebased - the branch is 987 commits behind the current `devel` branch (see https://github.com/bcoca/ansible/branches/all?page=2). The PR won't pass Ansible's test suite until the branch has been rebased.

mattclay commented:
CI failure in unit tests and integration tests, as well as timeouts on many integration tests.

gundalow commented:
!waffling stale_ci

bcoca committed:
Added ability to modules to emit 'update messsages'

  - new update_json method that sends 'intermediate messages'
  - new internal flag that can enable this facility from controller
  - method to run_command and log, log also requires debug to return data
  - 'live' keyword, enables 'live updates' from modules that support it
  - common method for connections to use fixes to module sending
  - spilt json filtering to return trailing data vs warnings
  - customized display
  - added 'live' toggles for adhoc/console
  - Fix performance degradation when running commands
    When all the file descriptors are empty and closed we still have to
    enter the if block to remove them from rpipes which means we will have
    to call `select()` again on the next iteration and as a result wait for
    the timeout unnecessarily which leads to a 1 second slowdown for each
    command being run.
  - Add live support to local connection

TODO:
  - Ouput layer
  - factor in async
  - only disable buffer if live
  - update stdin/stdout only if needed

bcoca committed:
fixes
